### PR TITLE
bpo-40204: Update documentation formatting for Sphinx 3.0

### PR DESCRIPTION
**Pull request by jakobandersen**
In the recently released Sphinx v3.0 the C domain has been rewritten to improve parsing of declarations and cross-referencing capabilities. However, this means that the existing directives has gotten more strict on the type of entities they accept.
This PR updates the documentation to use the new directives and roles. Many of the changes are related to C constructs in running text where a type or expression is "referenced". These have been changed to use the new markup role ``c:expr`` which accepts arbitrary types or expressions. This is in opposition to the other roles, like ``c:type`` which is for cross-referencing a type declared with a directive.

A few issues I couldn't immediately resovle:
- The refcount annotation in tools/extensions/c_annoations.py no longer works (though first broken in sphinx-doc/sphinx#7269).
- ``c-api/module.rst:194``: the old ``m_reload`` is technically declared inside the scope of ``m_slots``. This is now reflected in its ID/permalink.
- ``c-api/call.rst``: The two functions ``PyObject_CallFunctionObjArgs`` and ``PyObject_CallMethodObjArgs`` takes a variable amount of arguments, though the last one must be ``NULL``, indicated explicitly in the declaration. This is not valid C. There could be a configuration option in Sphinx to allow this. Please open an issue for Sphinx if you find this is relevant.
- ``howto/instrumentation.rst``: The ``c:function`` directive is used for some markers, but without a return type. The type ``void`` was added. However, ``python.function.return`` contains ``return`` which is a C keyword, so a warning is generated. The elegant solution would perhaps be to code a custom domain for these things, as they are not really C declarations (as far as I understand).
- ``c-api/unicode.rst``: The function ``PyUnicode_Translate`` is documented twice.
- ``c-api/init.rst``: There are several instances mentioning "``wchar_*`` string". Should this be ``:c:expr:`wchar_t*` string``?

https://bugs.python.org/issue40204

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

## Recognized GitHub username

We couldn't find a [bugs.python.org](https://bugs.python.org/) (b.p.o) account corresponding to the following GitHub usernames:

@jakobandersen

This might be simply due to a missing "GitHub Name" entry in one's b.p.o account settings. This is necessary for legal reasons before we can look at this contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.




You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

File: Doc/c-api/arg.rst [hunk 1]
```diff
@@ -59,9 +59,9 @@ which disallows mutable objects such as :class:`bytearray`.
    the length argument (int or :c:type:`Py_ssize_t`) is controlled by
    defining the macro :c:macro:`PY_SSIZE_T_CLEAN` before including
    :file:`Python.h`.  If the macro was defined, length is a
-   :c:type:`Py_ssize_t` rather than an :c:type:`int`. This behavior will change
+   :c:type:`Py_ssize_t` rather than an :c:expr:`int`. This behavior will change
```
Comment (serhiy-storchaka) [inline Doc/c-api/arg.rst:62]: Could it be possible to add the definition of the `:c:type:` role in the CPython specific extension?

Comment (jakobandersen) [reply Doc/c-api/arg.rst:62]: The ``:c:type:`` is still there, but it is (and was) meant for cross-referencing declared types, e.g., here for ``Py_ssize_t`` is is fine. Fundamental types like ``int`` and pointer types (e.g., ``Py_ssize_t*``)  can not be declared, so they can not be cross-referenced. They can still be typeset with relevant links, which is what the ``:c:expr:`` role is intended for.

Comment (serhiy-storchaka) [reply Doc/c-api/arg.rst:62]: Thank you for your clarification @jakobandersen. I hate to merge such large changes, but seems this is the only correct solution. Great work!

File: Doc/c-api/arg.rst [hunk 2]
```diff
@@ -331,13 +331,13 @@ Other objects
...
    takes two arguments: the first is a function, the second is the address of a C
-   variable (of arbitrary type), converted to :c:type:`void \*`.  The *converter*
+   variable (of arbitrary type), converted to :c:expr:`void*`.  The *converter*
    function in turn is called as follows::
 
       status = converter(object, address);
 
    where *object* is the Python object to be converted and *address* is the
-   :c:type:`void\*` argument that was passed to the :c:func:`PyArg_Parse\*` function.
+   :c:expr:`void*` argument that was passed to the ``PyArg_Parse*`` functions.
```
Comment (serhiy-storchaka) [inline Doc/c-api/arg.rst:340]: `:c:func:` not just changes the style and adds a link, but it also adds `()` after the function name. Please add explicit `()` in all sites where you changed `:c:func:` to `:c:expr:` or ``` `` `` ```.
```suggestion
   :c:expr:`void*` argument that was passed to the ``PyArg_Parse*()`` functions.
```

File: Doc/c-api/arg.rst [hunk 3]
```diff
@@ -652,9 +652,9 @@ Building values
 
    ``O&`` (object) [*converter*, *anything*]
       Convert *anything* to a Python object through a *converter* function.  The
-      function is called with *anything* (which should be compatible with :c:type:`void
-      \*`) as its argument and should return a "new" Python object, or ``NULL`` if an
-      error occurred.
+	  function is called with *anything* (which should be compatible with
+	  :c:expr:`void*`)as its argument and should return a "new" Python object, or
```
Comment (serhiy-storchaka) [inline Doc/c-api/arg.rst:656]:
```suggestion
	  :c:expr:`void*`) as its argument and should return a "new" Python object, or
```

File: Doc/c-api/arg.rst [hunk 4]
```diff
@@ -652,9 +652,9 @@ Building values
...
+	  ``NULL`` if an error occurred.
```
Comment (serhiy-storchaka) [inline Doc/c-api/arg.rst:657]: Seems the tab character is used here. Please replace all tab characters with spaces and check that lines are correctly aligned.

File: Doc/c-api/exceptions.rst [hunk 5]
```diff
@@ -189,7 +189,7 @@ For convenience, some of these functions will always return a
 .. c:function:: PyObject* PyErr_SetFromWindowsErr(int ierr)
 
    This is a convenience function to raise :exc:`WindowsError`. If called with
-   *ierr* of :c:data:`0`, the error code returned by a call to :c:func:`GetLastError`
+   *ierr* of :c:expr:`0`, the error code returned by a call to :c:func:`GetLastError`
```
Comment (serhiy-storchaka) [inline Doc/c-api/exceptions.rst:192]: I think it is enough to use
```suggestion
   *ierr* of ``0``, the error code returned by a call to :c:func:`GetLastError`
```
as for all literals.

**Review by vstinner — CHANGES_REQUESTED** [review]
The problem of this PR is that `:c:expr:` is not supported by Sphinx 2, and so this PR drops Sphinx 2 support. Since Sphinx 3 is not widely adopted yet, I proposed a different approach which keeps Sphinx 2 support, but adds Sphinx 3 support, see: https://bugs.python.org/issue40204 Sphinx 2+3 support can be added to stable branches, it's not a backward incompatible change. But my changes require Sphinx 3.2 which adds a new option to opt-in for Sphinx 2 syntax for the C domain.

This PR contains a few relevant changes which are compatible with Sphinx 2. If you want, please try to extract them and write a new PR.

**Comment by vstinner** [comment]
Thanks for your contribution! But your PR couldn't be merged. If you want, please try to extract your changes which are relevant and compatible with Sphinx 2.

**Pull request closed without merging**